File: Backend/Asistencia-con-Geofencing/src/main/java/ec/edu/espe/Asistencia_con_Geofencing/model/Attendance.java
package ec.edu.espe.Asistencia_con_Geofencing.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(
    name = "attendances",
    uniqueConstraints = @UniqueConstraint(columnNames = {"session_id", "student_id"})
)
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Attendance {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "session_id")
    private AttendanceSession session;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "student_id")
    private User student;

    @Column(name = "device_time", nullable = false)
    private LocalDateTime deviceTime;

    @CreationTimestamp
    @Column(name = "server_time", nullable = false, updatable = false)
    private LocalDateTime serverTime;

    @Column(precision = 10, scale = 8)
    private BigDecimal latitude;

    @Column(precision = 11, scale = 8)
    private BigDecimal longitude;

    @Column(name = "within_geofence", nullable = false)
    private Boolean withinGeofence;

    @Column(name = "sensor_status", length = 100)
    private String sensorStatus;

    @Column(name = "trust_score")
    private Integer trustScore;

    @Column(name = "is_synced", nullable = false)
    private Boolean isSynced = false;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "source_device_id")
    private Device sourceDevice;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "sync_batch_id")
    private SyncBatch syncBatch;

    @OneToMany(mappedBy = "attendance", cascade = CascadeType.ALL)
    private Set<SensorEvent> sensorEvents = new HashSet<>();
}

File: frontend/geofencing/lib/domain/entities/attendance.dart
import 'package:equatable/equatable.dart';

/// Entidad de Asistencia
class Attendance extends Equatable {
  final String attendanceId;
  final String sessionId;
  final String studentId;
  final String? studentName;
  final String? teacherName;
  final String? zoneName;
  final DateTime deviceTime;
  final DateTime serverTime;
  final bool withinGeofence;
  final double latitude;
  final double longitude;
  final String? sensorStatus;
  final int? trustScore;
  final bool isSynced;
  final String? syncDelay;
  final String? note;

  const Attendance({
    required this.attendanceId,
    required this.sessionId,
    required this.studentId,
    this.studentName,
    this.teacherName,
    this.zoneName,
    required this.deviceTime,
    required this.serverTime,
    required this.withinGeofence,
    required this.latitude,
    required this.longitude,
    this.sensorStatus,
    this.trustScore,
    required this.isSynced,
    this.syncDelay,
    this.note,
  });

  @override
  List<Object?> get props => [
        attendanceId,
        sessionId,
        studentId,
        studentName,
        teacherName,
        zoneName,
        deviceTime,
        serverTime,
        withinGeofence,
        latitude,
        longitude,
        sensorStatus,
        trustScore,
        isSynced,
        syncDelay,
        note,
      ];
}
